core/src/main/java/com/linecorp/armeria/common/metric/Metrics.java

```diff
@@ -0,0 +1,373 @@
+    }
+
+    /**
+     * Returns the {@link MetricGroup} with the specified key and type.
+     *
+     * @return the {@link MetricGroup} if found. {@code null} otherwise.
+     *
+     * @throws IllegalStateException if a {@link MetricGroup} with the specified key exists but its type
+     *                               mismatches the specified one.
+     */
+    @Nullable
+    public final <T extends MetricGroup> T group(MetricKey key, Class<T> type) {
+        requireNonNull(key, "key");
+        requireNonNull(type, "type");
+
+        final MetricGroup group = metricGroups.get(key);
+        if (group == null) {
+            return null;
+        }
+
+        if (!type.isInstance(group)) {
+            throw new IllegalStateException("metric group exists but its type mismatches: " + key);
+        }
+
+        @SuppressWarnings("unchecked")
+        final T castGroup = (T) group;
+        return castGroup;
+    }
+
+    /**
+     * Registers a {@link MetricGroup} into this registry or retrieves an existing one.
+     *
+     * @param key the {@link MetricKey} of the {@link MetricGroup}
+     * @param type the type of the {@link MetricGroup} produced by {@code factory}
+     * @param factory a factory that creates a {@link MetricGroup} of {@code type}
+     *
+     * @return the {@link MetricGroup} of {@code type}, which may or may not be created by
+     *         {@code factory}
+     * @throws IllegalStateException if there is already a {@link MetricGroup} of different class
+     *                               registered for the same {@link MetricKey}
+     */
+    public final <T extends MetricGroup> T group(MetricKey key, Class<T> type,
+                                                 BiFunction<Metrics, MetricKey, T> factory) {
+        requireNonNull(key, "key");
+        requireNonNull(type, "type");
+        requireNonNull(factory, "factory");
+
+        final MetricGroup group = metricGroups.computeIfAbsent(key, k -> factory.apply(this, k));
+        if (!type.isInstance(group)) {
+            throw new IllegalStateException(
+                    "A metric group of different type has been registered already for key: " + key +
+                    " (expected: " + type.getName() +
+                    ", actual: " + group.getClass().getName() + ')');
+        }
+
+        @SuppressWarnings("unchecked")
+        final T castGroup = (T) group;
+        return castGroup;
+    }
+
+    /**
+     * Returns the immutable {@link Collection} of {@link MetricGroup}s in this registry.
+     */
+    public final Collection<MetricGroup> groups() {
+        return Collections.unmodifiableCollection(metricGroups.values());
+    }
+
+    /**
+     * Returns the immutable {@link Collection} of {@link MetricGroup}s in this registry.
+     *
+     * @param type the type of {@link MetricGroup}
+     */
+    public final <T extends MetricGroup> Collection<T> groups(Class<T> type) {
+        return groupStream(type).collect(toImmutableList());
```

Comment:
It seems confusing that this method returns a copy while the other methods return an unmodifiable view. Should we make copies in all?